moveit_ros/perception/handeye_calibration_solver/src/handeye_solver_default.cpp

```diff
@@ -0,0 +1,373 @@
+        ss << *(double*)PyArray_GETPTR2(np_arg_eef_wrt_base[i], m, n) << " ";
+    }
+    ss << "\nnp_arg_obj_wrt_sensor";
+    for (size_t m = 0; m < ARRAY_SIZE; m++)
+    {
+      ss << "\n";
+      for (size_t n = 0; n < ARRAY_SIZE; n++)
+        ss << *(double*)PyArray_GETPTR2(np_arg_obj_wrt_sensor[i], m, n) << " ";
+    }
+    ROS_DEBUG_STREAM_NAMED(LOGNAME, ss.str());
+  }
+
+  // Import handeye.solver python module
+  p_name = PyString_FromString("handeye.solver");
+  p_module = PyImport_Import(p_name);
+  Py_DECREF(p_name);
+  if (!p_module)
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Failed to load python module: "
+                                        << "handeye.solver");
+    Py_DECREF(p_instance);
+    PyErr_Print();
+    return false;
+  }
+
+  // Find handeye.solver.solver_name class
+  p_class = PyObject_GetAttrString(p_module, solver_name.c_str());
+  Py_DECREF(p_module);
+  if (!p_class || !PyCallable_Check(p_class))
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Can't find \"" << solver_name << "\" python class");
+    Py_DECREF(p_instance);
+    PyErr_Print();
+    return false;
+  }
+
+  // Find handeye.calibrator.HandEyeCalibrator.solve method
+  p_func_solve = PyObject_GetAttrString(p_instance, "solve");
+  if (!p_func_solve || !PyCallable_Check(p_func_solve))
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Can't find 'solve' method");
+    Py_DECREF(p_class);
+    Py_DECREF(p_instance);
+    PyErr_Print();
+    return false;
+  }
+
+  // Create argument list for 'solve' method
+  p_args = Py_BuildValue("{s:O}", "method", p_class);
+  Py_DECREF(p_class);
+  if (!p_args)
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Can't create argument tuple for 'solve' method");
+    Py_DECREF(p_instance);
+    PyErr_Print();
+    return false;
+  }
+
+  // Call 'solve' method to solve AX=XB problem
+  p_value = PyEval_CallObjectWithKeywords(p_func_solve, nullptr, p_args);
+  Py_DECREF(p_args);
+  Py_DECREF(p_func_solve);
+  for (size_t i = 0; i < pose_num; ++i)
+    Py_DECREF(p_args_sample[i]);
+  Py_DECREF(p_instance);
+  if (!p_value)
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Error calling 'solve' method");
+    PyErr_Print();
+    return false;
+  }
+  PyArrayObject* np_ret = (PyArrayObject*)p_value;
+  if (!PyArray_Check(p_value) || PyArray_NDIM(np_ret) != 2 || PyArray_NBYTES(np_ret) != sizeof(double) * 16)
+  {
+    ROS_ERROR_STREAM_NAMED(LOGNAME, "Did not return a valid array");
+    Py_DECREF(p_value);
+    PyErr_Print();
+    return false;
+  }
+
+  std::stringstream ss;
+  ss << "\n Result camera-robot pose";
+  for (size_t m = 0; m < ARRAY_SIZE; m++)
+  {
+    ss << "\n";
+    for (size_t n = 0; n < ARRAY_SIZE; n++)
+    {
+      double item = *(double*)PyArray_GETPTR2(np_ret, m, n);
+      camera_robot_pose_(m, n) = item;
+      ss << item << " ";
+    }
+  }
+  ROS_DEBUG_STREAM_NAMED(LOGNAME, ss.str());
+
+  Py_DECREF(p_value);
+  ROS_DEBUG_STREAM_NAMED(LOGNAME, "Python C API end");
+  return true;
+}
```

Comment:
This solve() function is way too large and should be split into a few sub functions

Reply:
Since the embedded python usually requires to frequently use `Py_DECREF(variable)` to manually release a variable declared in a previous code. And some variables could be used much later, which may prevent a split in the middle. I will try to see if I can split this long function.

Reply:
I also recommend trying to split this function into smaller pieces.

Reply:
I read through this function, and it seems that it's big because it's interfacing with Python. The "right" way to fix it is probably to implement the whole thing in C++, but that's a rather significant undertaking. Maybe this PR could go ahead now, and the method could be ported in future work. (I would find this feature useful, and I'm willing to work on the C++ implementation.)

Reply:
The algorithm in the python library actually is the same with the one in [visp_hand2eye_calibration](http://wiki.ros.org/visp_hand2eye_calibration), which is a c++ library. But to my local test on a simulated robot arm with a fixed camera, and 50 random poses of the robot arm taken, the python library performs better than the c++ library on the final result. In addition, a new hand-eye calibration method has been added to OpenCV 3.4, see this [PR](https://github.com/opencv/opencv/pull/13880). But it will not compile against the default 3.2 version on Ubuntu 18.04. I haven't try this method yet, but it would be good if these libraries can be compared.